src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::context::{self, Context};
+use crate::events::EventNotifier;
+use crate::worker::{self, Worker};
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Order {
+    LeftToRight,
+    RightToLeft,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Strategy {
+    OneForOne,
+}
+
+#[derive(Debug)]
+pub struct StartError {
+    pub failing_child_runtime_name: String,
+    pub failing_child_error: Arc<worker::StartError>,
+    pub failing_sibling_termination_error: Option<TerminationError>,
+}
+
+#[derive(Debug)]
+pub struct TerminationError(HashMap<String, Arc<worker::TerminationError>>);
+
+pub struct Spec {
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+    pub children: Vec<worker::Spec>,
+}
+
+pub struct Supervisor {
+    pub runtime_name: String,
+    pub children: Vec<Worker>,
+
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+}
+
+fn ordered_children_on_start(
+    order: &Order,
+    children_specs: Vec<worker::Spec>,
+) -> Box<dyn Iterator<Item = worker::Spec>> {
+    match order {
+        Order::LeftToRight => Box::new(children_specs.into_iter()),
+        Order::RightToLeft => Box::new(children_specs.into_iter().rev()),
+    }
+}
+
+fn ordered_children_on_termination(
+    order: &Order,
+    children: Vec<Worker>,
+) -> Box<dyn Iterator<Item = Worker>> {
+    match order {
+        Order::LeftToRight => Box::new(children.into_iter().rev()),
+        Order::RightToLeft => Box::new(children.into_iter()),
+    }
+}
+
+// terminate_prev_siblings gets called when a child in a supervision tree fails
+// to start, and all it's previous siblings need to get terminated.
+async fn terminate_prev_siblings<'a, 'b, 'c>(
+    ev_notifier: &mut EventNotifier,
+    failed_child_spec: worker::Spec,
+    failed_child_start_err: Arc<worker::StartError>,
+    pending_children_specs: Vec<worker::Spec>,
+    started_children: Vec<Worker>,
+) -> (Vec<worker::Spec>, StartError) {
+    let prev_sibling_termination_result =
+        terminate_children(ev_notifier, started_children.into_iter().rev()).await;
+
+    let (prev_specs, failing_sibling_termination_error) = match prev_sibling_termination_result {
+        Err((prev_specs, sibling_err)) => (prev_specs, Some(sibling_err)),
+        Ok(prev_specs) => (prev_specs, None),
+    };
```

Comment:
Why not just return the type in this format originally, rather than doing this transformation here?

Reply:
A `Result<(), TerminationError>` is quite an awkward type to deal with, I think I'll leave this as is for now.